internal/alb/sg/security_group.go

```diff
@@ -0,0 +1,267 @@
+		controller.logger.Infof("revoking inbound permissions from securityGroup: %s", group.GroupID)
+		_, err := controller.ec2.RevokeSecurityGroupIngress(&ec2.RevokeSecurityGroupIngressInput{
+			GroupId:       group.GroupID,
+			IpPermissions: permissionsToRevoke,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to revoke inbound permissions due to %s", err.Error())
+		}
+	}
+
+	permissionsToGrant := diffIPPermissions(group.InboundPermissions, instance.IpPermissions)
+	if len(permissionsToGrant) != 0 {
+		controller.logger.Infof("granting inbound permissions to securityGroup: %s", group.GroupID)
+		_, err := controller.ec2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
+			GroupId:       group.GroupID,
+			IpPermissions: permissionsToGrant,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to grant inbound permissions due to:%s", err.Error())
+		}
+	}
+	return nil
+}
+
+// findExistingSGInstance tring to find the existing SG matches the specification
+func (controller *securityGroupController) findExistingSGInstance(group *SecurityGroup) (*ec2.SecurityGroup, error) {
+	if group.GroupID != nil {
+		instance, err := controller.ec2.GetSecurityGroupByID(*group.GroupID)
+		if err != nil {
+			return nil, err
+		}
+		if instance == nil {
+			return nil, fmt.Errorf("securityGroup %s doesn't exist", *group.GroupID)
+		}
+		return instance, nil
+	} else if group.GroupName != nil {
+		vpcID, err := controller.ec2.GetVPCID()
+		if err != nil {
+			return nil, err
+		}
+		instance, err := controller.ec2.GetSecurityGroupByName(*vpcID, *group.GroupName)
+		if err != nil {
+			return nil, err
+		}
+		return instance, nil
+	}
+	return nil, fmt.Errorf("Either GroupID or GroupName must be specified")
+}
+
+// diffIPPermissions calcutes set_difference as source - target
+func diffIPPermissions(source []*ec2.IpPermission, target []*ec2.IpPermission) (diffs []*ec2.IpPermission) {
+	for _, sPermission := range source {
+		containsInTarget := false
+		for _, tPermission := range target {
+			if ipPermissionEquals(sPermission, tPermission) {
+				containsInTarget = true
+				break
+			}
+		}
+		if containsInTarget == false {
+			diffs = append(diffs, sPermission)
+		}
+	}
+	return diffs
+}
+
+// ipPermissionEquals test whether two IPPermission instance are equals
+func ipPermissionEquals(source *ec2.IpPermission, target *ec2.IpPermission) bool {
+	if *source.IpProtocol != *target.IpProtocol {
+		return false
+	}
+	if *source.FromPort != *target.FromPort {
+		return false
+	}
+	if *source.ToPort != *target.ToPort {
+		return false
+	}
+	if len(diffIPRanges(source.IpRanges, target.IpRanges)) != 0 {
+		return false
+	}
+	if len(diffIPRanges(target.IpRanges, source.IpRanges)) != 0 {
+		return false
+	}
+	if len(diffUserIDGroupPairs(source.UserIdGroupPairs, target.UserIdGroupPairs)) != 0 {
+		return false
+	}
+	if len(diffUserIDGroupPairs(target.UserIdGroupPairs, source.UserIdGroupPairs)) != 0 {
+		return false
+	}
+
```

Comment:
Should we also compare `Ipv6Ranges`?

Reply:
it could be in future, if we used userIdGroupsPairs.. currently, i'd rather not write test cases for that 😄.
This refactor is just a start point to make code modular and ez to change, we can add ipv6 support easily in the future